### DISCUSSION TITLE
Store image from LDAP into pocketbase

### DISCUSSION THREAD

**Discussion by starredev** [pocketbase/pocketbase#3151] [Q&A] [answered]
I'm trying to store an base64 to an image in pocketbase.

unfortunatelyit is not store anything, do I anything wrong?

````
this.update = async (user) => {

    const formData = new FormData();

    formData.append('roles', user.roles)
    formData.append('title', user.title)
    formData.append('photo', Buffer.from(user.profilePhoto, 'binary').toString('base64'))

    const { result, error } = await asyncHandler(pocketbase.collection('users').update(user.pocketbase_user.result.id, { formData }))    
    
    if(error) {
        throw 'Cannot update user. If error keeps returning please raise an incident'
    }

    return result;
}
````

**Comment by ganigeorgiev** [comment]
You have to pass a [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) instance.

In your case, if you have some node buffer instance, doing `new Blob([buff])` should suffice.

**Reply by starredev** [reply]
Hi,

I tried both, to store it as base64 or as binary ```formData.append('photo', new Blob([user.profilePhoto], {type: 'jpeg'}))``` like this way. It doesn't save the image

**Reply by ganigeorgiev** [reply]
What is `user.profilePhoto` in the above?

You can also start the executable with `--debug` to see if there are any logs in the console.

**Reply by ganigeorgiev** [reply]
Oh, I just also noticed that you are not submitting the FormData properly.

In your code:
```js
update(user.pocketbase_user.result.id, { formData })
```
Should be:
```js
update(user.pocketbase_user.result.id, formData)
```

_(aka. without the object wrapping)_

**Reply by starredev** [reply]
That makes sense!

Now its storing an txt with base64, but its saving now though.

userProfile is binary from ldap.

````
/9j/4AAQSkZJRgABAQEAYABgAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCABgAFYDASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwDg7/UL+LUrg/bbjHntn94em48dajTU9QlwReznBxxIcEfnV2/soo9TuGnnAXzWJXPXmoXvNPtcFEDMBgEdqaAW3udTmG9Z7rzFztxIwA/WtCNNQzul1CdR3AlPP61jza+7jbGoUg4z3qk99dzj53IHUZNUk2Fzq/t6Wy4N7O23n5pSc1Xl8WzR/LFNLj1LmuWI5O6XPOevSlyibtqdsEnvQo63ZN0alx4p1WT7l1Og7neeKypdZ1R33tqF1ye8pH9aqyszPheQx4wP0o+xgJvdhwcY61VrDRMNXvw3N/c5/wCurf41ImsX5yVvLnHqZW/xrNMePuHOKcuVxhhjPA70ldjZsLqOpSDJ1G6HpiU80VWjdHX5iGx6HH40Vpyt7GN2WNXMr6neb5DxNJ3/ANo1UKqSAQW4/hqbVpgurXpEa/69+oz/ABGqbytj72B6DtUbampYK7VBQbMnBwOacF4ALFiB0PSnqD9nRs5GOBjpTv8AVqzDpVJGcmNAJxtB9+OlNljBU5YjP6U153H8OMjjNQASTyLErHe7BQMc5NJysCiy/pek3F9loIy3ZSeOPWuhj8NaitttFtk/w8jOa6rw3oAt7OMEYYgD3rrU08CPpmuSVZ30O2FDS7PE7/w/qsSF2sn4OMqQc/hWNNaSQhGkXbuOSp4Ne73llhTmuE8T6YkqKyookUn5scn2rSjXvKzHUoWjdHnyBslRkkdQO1FWLiBre4KMCxA7daK7HFJnJcNU51a+6f69+v8AvGqxJwpbJA4zVrVMDV70EDHnuSfQ7jVU4OQcHI5rHQZo26D7CSeq4I57Zx+NPGdpCkYxzmm2i7rT5TypPBqUYwcDp0qlbqY1GUcMpxyMfjV/Qo4Trtq05Kqp3evP+f5VTbl8HJznPFdH4MgLXdzcLCHljQBFYdc//qrOppFs3prmaPU9BntbtA1vMkgB/hNdC0ISMu7BVAycnFcrDaGK7WZYo4lwhEkeRvY4yuPz5rcvfLu5hCxyNvQng1xaHendGffXtquQJVb6HNc1eJFelyjglf0rclje1lngj0yIRxkbZkwfMB7jI7VTktEQF1RVZuuB1rahGPtEpClK8TyjxLDt1UA94/60V2UngtNauvtVxcSJlcKsYHAz3z9aK7q9WmqjS6HJGhOSuee6qP8AidXgHP758f8AfRqtksuDjk4zirmprnV77gZ89yOenzGovIJjD7kAIyBu56+n4Vj5szLljn7MwwCQTU/y4JwOR2qPTcKkgU8BuKkUA4yOR1q42sY1Cpj5sHK/McYPSuw8AADV5oD0kQN164rk3z5pGCOeDXQ+D7hbbXo2zkshAx3rKrrFo3ovVM9Te2lbVYIsMyxjzD6D0q/5Qa4VWBBJwCRVa3urmW8EgdUGPXn2q4sjJL5m9WwM43ZINcSWh6HMJfWrLHl2yB0rmryTYrMO1dLqN0r2+FOOOa4PXtSjs4vNfkAjC56muzBKPtbsipL3Czqerpo2n2uEImOFKn6ZJ/lRXH3GpS6xdNPPtTjCoDwoorS6Od17aI5rUudXvh1xO+Pf5jU66PfSWf2oWs32dY2cvswuAcE5PUAkD8aranldYvMDP798f99GlVpZlUSzEoB8qFs470m3bQxLumQFt5BxnoCfTrWlc6bDbWn2gXUTkYOxTzzWdFdjylQJt2ZO7aBnPqRyfxqq19lQrk470oX6kVEPulAuWVQAoYcn19altZTZXscsUyuYzuDL0qisqSPIzMzMTxj9M1CtxGhGVY9cLmiV+hUdEe36DqsOp2ayxFfM29CM5ra3LHGJLiOMMB0Va870BbvS7KGSEASbcujdK2Lm+1W/Vf3KLjod/Fcjir2R3wnLlL+o6xCEcIMKOSc9q4LxLdRvFBNcCRY3c+UF4zgdT7V0dtokjymW9maTnOwcLWP43tVk0kyghTbsHH8iP1rqwvKpNX1sY1VJxOdt9RjhDbNvzdS3JornUkOTlQfxxRWvu9TmsauogjVr3bHnNw/Pp8xqe1jQxKWRD3PzHPvVTVXkGpXwR2x5749/mNVHdgoOX6YGDUPVDNTVHgjgTyEMchbkk/piscyyEBffsM5pwXg7gxI9TShAAMA57Yo2AYU2HLZyewPH40BnDB9vK4x+FOZSTnOPTFKEBABAPbgUncZ7DpLJd2cF1HgpKgYY962I1VRtxivIdA8UXuhL5MarPa53eU5PB9VPb6V1o+I9mUVpNNuEbvhlIFcsqU09DrhVjbU7GcbIsDAzXnHjjVY226VE27ndcMO2Oi/XvTNY8fXt6GjsYfsseMGRjuf8OwrjmLSHJ3ZY5JzksfUmrp0ne8iKtVNWiM7nHPbpRSqrNnGfworpT8jlP//ZAA==
````

**Reply by starredev** [reply]
So the code below let me save me the photo, how could I save it in pocketbase?

````
    require("fs").writeFile("test.jpeg", user.profilePhoto, 'base64', function(err) {
        console.log(err);
    });
`````

I tried serveral things without luck.

The code below stores the image but it seems corrupt with no image

````ormData.append('photo', new Blob([user.profilePhoto], { type: 'image/jpeg' }), 'user.jpeg')````

**Reply by ganigeorgiev** [reply]
@raizdev I don't recommend the above, but I'm also failing to understand what you are trying to do. Why do you need the file in base64 in the first place?

The only thing you have to do is to create a `Blob` from the file content/binary data.

**Reply by ganigeorgiev** [reply]
If you have a base64 string and want to convert it to blob, you can check https://stackoverflow.com/questions/16245767/creating-a-blob-from-a-base64-string-in-javascript.

You can also try with something like:

```js
new File([Buffer.from("yourBase64String", "base64")], "test.jpg", { type: "image/jpeg" });
```

But in any case, this is not specific to PocketBase. The stored image is corrupted/invalid in your case because you are filling it with invalid data.

**Comment by starredev** [comment]
> Oh, I just also noticed that you are not submitting the FormData properly.
> 
> In your code:
> 
> ```js
> update(user.pocketbase_user.result.id, { formData })
> ```
> 
> Should be:
> 
> ```js
> update(user.pocketbase_user.result.id, formData)
> ```
> 
> _(aka. without the object wrapping)_



> @raizdev I don't recommend the above, but I'm also failing to understand what you are trying to do. Why do you need the file in base64 in the first place?
> 
> The only thing you have to do is to create a `Blob` from the file content/binary data.

I suppose that A blob isn't a valid image source. I have to convert the binary data to base64 so I can pass it as an image source. In ldap I could take 'thumbnailPhoto;binary' attribute it throws me an base64. Now i'm trying to save it in pocketbase. I also tried to store directly the binary file to pocketbase which doesn't work either. Im quite new to this.

If I pass the binary directly into the blob it stores a file blob_34jdj3h into pocketbase which isnt a image

**Reply by ganigeorgiev** [reply, marked as answer]
I think you have some misunderstanding with the data you are operating and whether it is binary or base64, but I'm not sure how to help you with that.

Here is an example with base64 image:

```js
import PocketBase from "pocketbase"

const pb = new PocketBase("http://127.0.0.1:8090")

const base64Img = "iVBORw0KGgoAAAANSUhEUgAAASwAAAEsBAMAAACLU5NGAAAAG1BMVEXMzMyWlpacnJyqqqrFxcWxsbGjo6O3t7e+vr6He3KoAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAD90lEQVR4nO3cwW+bSBQH4AcGw5HnJDhHaN3dHO1su9ojNGnPtrUb7dFuIiVHnEo5263Uv3vfGwab1myVA5DV6vcpgeD35HmeGYbJxUQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD/zOb3N5BRexlu9/Jo+NCQFl/HbWrRK7s6Amcdy3jCfaftyOT/OmsnLxSFqkzu04Ns1Z+RxPOMtUc63fH6U5HP8O5/uo1Vyh9IJhTylwSjz0pV0y4Tex0dJ7iij3ck+WiV3J9RPvVhRLgO5O5V+KOSl7MesnXSRH++jNrlDAWurEW0i6ZOz8jI9mlwaDXkftckd8nXEdgnNVjI2sf6Q+VvLSMiMHJnupHC0j9rkrmlL87Lhs7JK86oM1fowVFq0jdrkjn2QKbMuTEvD8aGsfCQ9th9PbzHeR21yt1KWkUq3et+Tq4tDHpnXfZ67+7Zdltu1itrkbrEuRWVLWdmwHbl0shlXSQ7LLVtFbXLXZUmLphHOHK3IsWVtTg6Lk6PFV1Gb3G1Z9I1Xjb015NpSHq7jfntL7reoaW7JhD+pJQ2537llVuyGO1Em17iWJMt7f3ei/zeZcdGlKLDr1saW5XPV9F9bM2pV1CZ3yDxDZFx0HZcF0z+s8rpwVcuWPo5k1KqoTe7QwD58mp6Js/PUTn4tVEatx2ei3lAzu4M4t3uErQl5PN3YOb84NR+gitrkDnl8J51QNO23hjLH7SqQxxnp0trbfotmo9t0RE27U9k9hFw2PuBfLnVD0d/u9KMs8hNq2svrxFqXJXprZtmg9riXp5v0jTRI4afyn5lv1X8+gRaQ22XA/zT6sxatkgEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD65tjf/5gXLYtHb/8l8kNZkVw5zEwjIjei8ru7rtJ7/YqcO3ISorTFsiLvt+eXZY7xlp5sWd6b7KscrpeZ80DBus2y6D1dviY3C+QP/9WUnGWkp8GrhZa1fE3DQiK1ssYrurdlDeblwZ86TzTctFuWf/dxPihy+kw31+/IuTOnm2v98I6EwoTe1cuKLsLEluVm5cFLHHf7pc2JKIPoZl4STpfFHzSRfnEyc5pQrmVJiO7l13yRHpdlPQ0LW5ZTHSInWN23WZZMedMJycUq0aa1FT1F1dyK6MugoHpvuY903Fv0a9Jqb+n7apesHlY0KSvRU6233CV9V5Z/RsdzixbzlsvSuUXL4nFOT9mVtq2nw9yiYPx9WebCHGt3IrW7yOnby51IuyzPKEgv9M31dLgTKUgayioH+oqrdavlsp5hWPTb3jM9vnQBjZyLl64AAP43/gHVSaMe2vmdiAAAAABJRU5ErkJggg=="

const formData = new FormData()
formData.append("myImage", new Blob([Buffer.from(base64Img, "base64")]), "test.jpg")

await pb.collection("example").create(formData)
```

(_I'm using `Blob` because `File` doesn't seem to be available in Node.js 18 by defualt_)

**Reply by starredev** [reply]
Wow.. that works.. I feel so sorry, thankyou so much for your help.
I had In buffer.from option binary instead of base64 and I didn't see it.

**Comment by starredev** [comment]
@ganigeorgiev db.files throws me undefined while db.collection and db.admins works fine. 
Any issues you know around it?

**Reply by ganigeorgiev** [reply]
I'm not sure I follow. What is `db.files`?
Is `db` a JS SDK PocketBase instance? If Yes, what version of the JS SDK do you use and what call to the files service do you make?

**Reply by starredev** [reply]
0.10.0, updated to 0.16.0 works fine now. Thanks!
